lighthouse-core/audits/polyfills.js

```diff
@@ -0,0 +1,233 @@
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class Polyfills extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'polyfills',
+      scoreDisplayMode: Audit.SCORING_MODES.MANUAL,
+      description: str_(UIStrings.description),
+      title: str_(UIStrings.title),
+      requiredArtifacts: ['Scripts'],
+    };
+  }
+
+  /**
+   * @param {Poly[]} polyfills
+   * @param {string} code
+   * @return {PolyIssue[]}
+   */
+  static detectPolyfills(polyfills, code) {
+    const qt = (/** @type {string} */ token) =>
+      `['"]${token}['"]`; // don't worry about matching string delims
+    const pattern = polyfills.map(({object, property}) => {
+      let subpattern = '';
+
+      // String.prototype.startsWith =
+      subpattern += `${object || ''}\\.?${property}\\s*=`;
+      subpattern += `|${object || ''}\\[${qt(property)}\\]\\s*=`;
+
+      // Object.defineProperty(String.prototype, 'startsWith'
+      subpattern += `|defineProperty\\(${object || qt('window')},\\s*${qt(property)}`;
+
+      return `(${subpattern})`;
+    }).join('|');
+
+    const polysSeen = new Set();
+    const polyMatches = [];
+    const re = new RegExp(`(^\r\n|\r|\n)|${pattern}`, 'g');
+    /** @type {RegExpExecArray | null} */
+    let result;
+    let row = 0;
+    let rowBeginsAtIndex = 0;
+    // each poly maps to one subgroup in the generated regex. for each iteration of the regex exec,
+    // only one subgroup will be defined. Exec until no more matches.
+    while ((result = re.exec(code)) !== null) {
+      // discard first (it's the whole matching pattern)
+      // index 1 is truthy if matching a newline, and is used to track the row number
+      // matches maps to each possible poly. Exec until no more matches
+      // only one of [isNewline, ...matches] is ever defined.
+      const [, isNewline, ...matches] = result;
+      if (isNewline) {
+        row++;
+        rowBeginsAtIndex = result.index;
+        continue;
+      }
+      const poly = polyfills[matches.findIndex(Boolean)];
+
+      // don't report more than one instance of a poly for this code.
+      // would be nice, but it also reports false positives if both '='
+      // and 'Object.defineProperty' are used conditionally based on feature detection.
+      if (polysSeen.has(poly)) {
+        continue;
+      }
+
+      polysSeen.add(poly);
+      polyMatches.push({
+        row,
+        col: result.index - rowBeginsAtIndex,
+        poly,
+      });
+    }
+
+    return polyMatches;
+  }
+
+  /**
+   * @return {Poly[]}
+   */
+  static getPolyDefinitions() {
+    // if latest Chrome supports a feature natively, we should
+    // complain about the existence of a polyfill
+    return [
```

Comment:
note to self: try replacing this with some logic. look for any `<something?>.prototype.<something>`, if pattern is found ensure it is a native built in (does it exist?) and not some custom thing.